---
title: "PhyloguidesR Tutorial"
author: "Holly K. Arnold"
date: "2025-03-25"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## PhyloguidesR  
The PhyloguidesR package is a package built to help researchers build the most accurate trees from short-read microbial sequencing data to enable subsequent phylogenetic microbial community analyses (e.g. unifrac or Cladal Taxonomic Unit analyses).

The 16S rRNA gene has been key to sequence-based phylogenetic microbial community analyses for over 30 years. Since the full length 16S gene (~1500 base pairs) is longer than what short-read sequencing technologies can typically capture, researchers must target a portion of the gene, or variable region, typically ~250 base pairs in length. 

Prior work has shown that the accuracy of phylogenetic trees are improved when they are first aligned with full length guide sequences (see Figure right) rather than using the short reads alone (see Figure left). 

<div style="text-align: center;">
  <img src="inst/images/GuideUsePicture.png" alt="Descriptive alt text" width="600"/>
</div>

This makes sense given what we know about the full length 16S gene. It is likely that the full lengh sequences provide additional phylogenetic context to build more accurate trees as well as serve as a scaffolding for mapping of short reads. 

This publication is currently in the review processes, and I will link to it when it becomes available.

For the purposes of this tutorial, we will show how to perform the basic workflow listed on the right side of the figure above to make the most accurate phylgoenetic trees using guide sequences for subsequent use of phylogenetic microbial community analyses (e.g. Unifrac, or Cladal Taxonomic Unit Analysis). 

We assume that the user has a phyloseq object and that they are wanting to build a phylogenetic tree from "short read" ASVs. 

## Installing PhyloguidesR

You should be able to install using `remotes::install_git("phyloguidesR")`.

## Step 0: Prepare data for the tutorial
We will use a phyloseq object that is example of microbial short read data. The data that we will work with was derived from [mothur MiSeq SOP](https://mothur.org/wiki/miseq_sop/), where raw reads were processed to generate ASVs via the [dada2 1.16 tutorial pipeline](https://benjjneb.github.io/dada2/tutorial.html). The phyloseq that you will load was generated by 2x250 Illumina Miseq amplicon sequencing of the V4 region of the 16S rRNA gene from gut samples collected longitudinally from a mouse post-weaning. For learning purposes, we also included an example "bad ASV" which we will
filter out later. 

For now, you can just consider them as an example of microbial short read data. Our goal is to build an accurate phylogenetic tree from this short read data with use of full length guides.

```{r, eval = FALSE}
library(phyloguidesR)
library(phyloseq)
data("ex_ps")
ex_ps
```

```{r, eval = TRUE, echo = FALSE}
suppressWarnings(library(phyloseq))
library(phyloguidesR)
data("ex_ps")
ex_ps
```

As you can see, this phyloseq object contains an (1) OTU Table (aka ASV table), (2) some sample data, and (3) a tax table, but it doesn't contain a phylogenetic tree (which would be notated as phy_tree) - this is what we are trying to build. If you're unfamiliar with phyloseq objects, you should familiarize yourself more with them [here](https://vaulot.github.io/tutorials/Phyloseq_tutorial.html). In brief, you can access each of the components of the phyloseq object by running the following commands: 

```{r, eval = FALSE}
otu_table(ex_ps) # Access the OTU table
tax_table(ex_ps) # Access the TAX table
sample_data(ex_ps) # Access the SAMPLE DATA table
```


For the purposes of our tutorial, we will be using guide sequences that were first derived from the [Silva Living Tree Project](https://www.arb-silva.de/silva-license-information/) which were further curated to an approximated [Silva Seed Reference 138.2](https://mothur.org/wiki/silva_reference_files/) for pratical use . The latest SILVA release is under the CC-BY license. 

We maintain the phyloguidesR-compatible [Seed Silva Reference Files Release 138.2](https://mothur.org/wiki/silva_reference_files/). If you want to follow along for the tutorial, download the file `guides_silva_138_2_seed.fasta` and `guides_silva_138_2_seed.align` accessed for download [here](https://zenodo.org/records/15086268). 

<div style="border-left: 5px solid #4CAF50; background-color: #eaf6ea; padding: 1em; margin: 1em 0;">
<strong>Suggestions for your own data</strong>
<ul>
  <li>We provided properly formatted curated guide sequences as described above, but you can customize to use any set of guides that you find to be most appropriate for your microbial community of interest.</li>
  <li>To customize your full length 16S guide sequences, provide them formatted as a FASTA.</li>
  <li>Guide sequences and ASVs will be aligned to a template alignment file. We provided a properly formated curated alignment file as described above</li>
   <li>To customize your full length 16S alignment template file, provide them formatted as multi-aligment FASTA.</li>
</ul>
</div>

Last, set the directory path (DIR.GUIDES) to where you stored your full length guides (`guides_silva_138_2_seed.fasta`) and template alignment file (`guides_silva_138_2_seed.align`), as well the path to where you want output files to be written (DIR.OUT).

```{r, eval = FALSE, echo = TRUE}
# Set to a directory where you want to look at output. 
DIR.OUT = "/path/to/desired/output/folder/"
DIR.GUIDES = "/path/to/where/you/stored/guides/fasta/file/"
```

```{r, eval = TRUE, echo = FALSE}
DIR.OUT = "/Users/arnoldhk/Desktop/Research/2025_HVR_Guide_Phylogenetic_Integration/phyloguidesR/phyloguides_tutorial_data/out/"
DIR.GUIDES = "/Users/arnoldhk/Desktop/Research/2025_HVR_Guide_Phylogenetic_Integration/phyloguidesR/phyloguides_tutorial_data/data/"
```


## Step 1: Combine your ASVs from a phyloseq with full length guides.
As a first step, we write out our ASVs in fasta file format. 
```{r, eval = TRUE, echo = FALSE}
# Write out the fasta file
asvs = file.path(DIR.OUT, "asvs.fasta")
```

```{r, eval = FALSE, echo = TRUE}
# Write out the fasta file
asvs = phyloseq_to_fasta(phyloseq = ex_ps, dir_out = DIR.OUT, filename_out = "asvs.fasta")
```

```{r, eval = FALSE, echo = TRUE}
# ✅ Show first liness of the fasta file - make sure they are what you expect
first_lines <- system(my_cat(list = c("head  ", asvs)), intern = TRUE)
cat(first_lines, sep = "\n")

# ✅ Show last lines of the fasta file - make sure they are what you expect. 
last_lines <- system(my_cat(list = c("tail  ", asvs)), intern = TRUE)
cat(last_lines, sep = "\n")
```

Then, combine ASVs with guide sequences in a fasta file. 
```{r, eval = FALSE, echo = TRUE}
guides_and_sequences = add_guides(dir_out = DIR.OUT, 
                                  filename_out = "seqs_and_guides.fasta", 
                                  guides_file_path = file.path(DIR.GUIDES, "guides_silva_138_2_seed.fasta"), 
                                  asvs_file_path = asvs)
```

```{r, eval = TRUE, echo = FALSE}
guides_and_sequences = file.path(DIR.OUT, "seqs_and_guides.fasta")
```

```{r, eval = FALSE, echo = TRUE}
# ✅ Show first liness of the fasta file - make sure they are what you expect
first_lines <- system(my_cat(list = c("head  ", guides_and_sequences)), intern = TRUE)
cat(first_lines, sep = "\n")

# ✅ Show last lines of the fasta file - make sure they are what you expect.
last_lines <- system(my_cat(list = c("tail  ", guides_and_sequences)), intern = TRUE)
cat(last_lines, sep = "\n")
```

Notice we should see full length guide sequences at the tail of our combined file output now. Each reference sequence has a pattern `REF.` in front of it, and you may also notice that the guide sequences are quite a bit longer compared to our sequences amplified from the V4 region. 

## Step 2: Align your ASVs + guides 
In order to build a phylogenetic tree, you need to build a high quality Multiple Sequence Alignment file. To do so, we will use [mothur's implementation of NAST algorithm](https://mothur.org/wiki/align.seqs/). Note that phyloguidesR provides a wrapper for the `align.seqs` function for convience purposes, but there is no reason that you couldn't just run that function straight from the command line as well. The NAST algoirthm works by aligning a candidate sequence to a template alignment. The high quality alignment provided above was derived from a [recreation of the SILVA SEED database](https://mothur.org/wiki/alignment_database/) and subsequently formatted for use with the phyloguidesR package. 

Lets go ahead and run the mothur align.seqs command with the phyloguidesR wrapper. As you can see, we pass the candidate file that we want to align - the fasta file containing our ASVs with full length reference sequences. We align that to a gold standard template alignment file. The output directory says where to store the alignment file, and last, the processors says how many cores one should use. 

```{r, eval = FALSE, echo = TRUE}
# Align guides and sequences
guides_and_seqs_align = align_seqs_mothur(candidate = guides_and_sequences,
                                          output.directory = DIR.OUT,
                                          template = file.path(DIR.GUIDES, "guides_silva_138_2_seed.align"),
                                          processors = 1)
print(basename(guides_and_seqs_align))

```

```{r, eval = TRUE, echo = FALSE}
guides_and_seqs_align = file.path(DIR.OUT, "seqs_and_guides.align")
print(basename(guides_and_seqs_align))

```

<div style="border-left: 5px solid #f44336; background-color: #fdecea; padding: 1em; margin: 1em 0;">
<strong>Pitfalls and Suggestions</strong>
<ul>
  <li>If mothur is not installed on your machine, install it [here](https://github.com/mothur/mothur/releases/tag/v1.48.2)</li>
  <li>You will need to add mothur to your PATH variable. Call `which mothur` in your terminal to see if you have successfully added it to your path.</li>
  <li>If you have a mac, you will then need to then allow mothur to run. Try running mothur once first by typing `mothur` in the command line - you'll see the "unidentified developer message". Open System Preferences > Security and Privacy > General and then you'll see "mothur was blocked from use because it is not from an identified developer." Click "Allow Anyway". </li>
  <li>Additional documentaiton is located [here](https://mothur.org/wiki/align.seqs/) </li>
</ul>
</div>

Ok, now lets take a look at outputs. You should see that there are gaps (- or .) which have been inserted into the sequences.
```{r, eval = FALSE, echo = TRUE}
# Look at the alignment file produced
first_lines <- system(my_cat(list = c("head  ", guides_and_seqs_align)), intern = TRUE)
cat(first_lines, sep = "\n")

# Look at the alignment file produced
first_lines <- system(my_cat(list = c("tail  ", guides_and_seqs_align)), intern = TRUE)
cat(first_lines, sep = "\n")

# Look at the log file
logfile <- system(my_cat(list = c("cat  ", DIR.OUT, "*.logfile")), intern = TRUE)
cat(logfile, sep = "\n")
```

## Step 3: Screen your sequences to remove those with low qualtiy information.
Next we will remove any sequences from the alignment which don't have at least a minimum number of bases. Such sequences are likely not going to provide enough information to build high quality trees. This will be done using mothur's `screen.seqs` command. For convience purposes, we provided a wrapper for this function `screen_seqs_mothur` but there is no reason you couldn't just run this in command line yourself. 

```{r, eval = FALSE, echo = TRUE}
# Screen for sequences which are poorly aligned (e.g. have less than 100 characters of information)
guides_and_seqs_align_screened = screen_seqs_mothur(align_path = guides_and_seqs_align, min_seq_length = 100)
print(basename(guides_and_seqs_align_screened))
```

```{r, eval = TRUE, echo = FALSE}
# Load the premade alignment file
guides_and_seqs_align_screened = sub(guides_and_seqs_align, pattern = ".align$", replacement = ".good.align")
print(basename(guides_and_seqs_align_screened))
```

Now take a look at the sequences that did not meet our requirement of at least 100 bases in the alignment.
```{r, eval = FALSE, echo = TRUE}
# Sequences which did not aling and have been filtered from the .good.align file
bad_sequences = sub(guides_and_seqs_align_screened, pattern = ".good.align$", replacement = ".bad.accnos")
bad_sequences <- system(my_cat(list = c("head  ", bad_sequences)), intern = TRUE)
cat(first_lines, sep = "\n")
```

<div style="border-left: 5px solid #4CAF50; background-color: #eaf6ea; padding: 1em; margin: 1em 0;">
<strong>Suggestions for your own data</strong>
<ul>
  <li>Explore each of the output files in command line using the `head <filename>` command.</li>
</ul>
</div>


<div style="border-left: 5px solid #f44336; background-color: #fdecea; padding: 1em; margin: 1em 0;">
<strong>Pitfalls and Suggestions</strong>
<ul>
  <li>You can also try running this step on the command line if its easier. Additional documentaiton is located [here](https://mothur.org/wiki/screen.seqs/). </li>
</ul>
</div>

## Step 4: Filter alignment for gaps. 
Last, there are some columns which likely just contain gaps (- or .)  which means those columns essentially hold no information about the alignment. You should filter these out. We provide a wrapper function for mothur's `filter.seqs` command, but there is no reason you couldn't run this on command line yourself. 

```{r, eval = FALSE, echo = TRUE}
# Filter alignment for gaps
guides_and_seqs_align_screened_filtered = filter_seqs_mothur(fasta_path = guides_and_seqs_align_screened)
print(basename(guides_and_seqs_align_screened_filtered))
```

```{r, eval = TRUE, echo = FALSE}
# Load the premade filtered file
guides_and_seqs_align_screened_filtered =  sub(guides_and_seqs_align_screened,
                                               pattern = ".align$",
                                               replacement = ".filter.fasta")
print(basename(guides_and_seqs_align_screened_filtered))
```
<div style="border-left: 5px solid #4CAF50; background-color: #eaf6ea; padding: 1em; margin: 1em 0;">
<strong>Suggestions for your own data</strong>
<ul>
  <li>Explore each of the output files in command line using the `head <filename>` command.</li>
</ul>
</div>

<div style="border-left: 5px solid #f44336; background-color: #fdecea; padding: 1em; margin: 1em 0;">
<strong>Pitfalls and Suggestions</strong>
<ul>
  <li>You can also try running this step on the command line if its easier. Additional documentaiton is located [here](https://mothur.org/wiki/filter.seqs/) </li>
</ul>
</div>


## Setp 5: Make a phylogenetic tree
We now have made a curated multi-sequence alignment file which has been filtered to remove low quality alignments, and columns without any information. We are finally ready to make a tree form this alignment file! Let's run FastTree - this step can take some time for the tree to be made. 

```{r, eval = FALSE, echo = TRUE}
# Build tree
tree_path = make_fasttree(path_alignment = guides_and_seqs_align_screened_filtered,
                          fasttree_flags = "-nt -gtr -gamma",
                          output_file_name = "tree_with_guides.tre",
                          output_dir = DIR.OUT)
print(basename(tree_path))
```

```{r, eval = TRUE, echo = FALSE}
# Load the pre-made filtered file
tree_path =  file.path(DIR.OUT, "tree_with_guides.tre")
print(basename(tree_path))
```


<div style="border-left: 5px solid #4CAF50; background-color: #eaf6ea; padding: 1em; margin: 1em 0;">
<strong>Suggestions for your own data</strong>
<ul>
  <li>Explore each of the output files in command line using the `head <filename>` command.</li>
  <li>It may be easier to run this command directly on command line. That's fine too! You can find information about how to run on command line [here](https://morgannprice.github.io/fasttree/#Usage) </li>
</ul>
</div>

<div style="border-left: 5px solid #f44336; background-color: #fdecea; padding: 1em; margin: 1em 0;">
<strong>Pitfalls and Suggestions</strong>
<ul>
  <li>More information on FastTree and the flags we used is found here [here](https://morgannprice.github.io/fasttree/#Usage) </li>
   <li>If FastTree isn't installed on your system, you'll have to download it here [here](https://morgannprice.github.io/fasttree/#Install) </li>
</ul>
</div>

## Step 6: Midpoint Root 
Well, now that we've made a phylogenetic tree, we are ready to import the tree back into R. If subsequent analyses rely on a rooted tree, a common method chosen when the root isn't known a-priori is  [midpoint rooting](https://www.ebi.ac.uk/training/online/courses/introduction-to-phylogenetics/what-is-a-phylogeny/aspects-of-phylogenies/root/). In the case of midpoint rooting we simply choose the root as the midpoint between the two longest branches. An assumption of this is that all of our sequences are evolving at the same rate, which is likely more appropriate for the 16S gene as compared to genes undering fast varied rates of evolution.

```{r, eval = TRUE, echo = TRUE}
# Build tree
tree = ape::read.tree(tree_path)
print(basename(tree_path))
tree

#Midpoint root
midpoint.tree = phangorn::midpoint(tree)
midpoint.tree
```

## Step 7: Drop guide sequences
Last, while guides have been helpful to improve the accuracy of this tree, we will need to drop out reference sequences to perform our microbial community analyses.

```{r, eval = TRUE, echo = TRUE}
# Get a list of reference tip names
ref.tips <- grep("^REF\\.", midpoint.tree$tip.label, value = TRUE)

# Drop those tips
pruned_tree <- ape::drop.tip(midpoint.tree, ref.tips)
pruned_tree
```

## Step 8: Add the phylogenetic tree to your phyloseq object. 
Last, we need combine our tree with our phyloseq object. Remember, that our phyloseq has ASV names (e.g. ATTGGGTT...), while it looks like (above) our fasttree has human readable names (e.g. seq1, seq2, ...). We can find a mapping of the human readable name to the sequence in our initial fasta file we wrote out.

```{r, eval = TRUE, echo = TRUE}
# Map our sequence names to the FASTA sequence
key_to_seq_map = phylotools::read.fasta(asvs)
head(key_to_seq_map)

# Rename our phyloseq object with the human readable names.
ps_renamed = rename_taxa_with_short_names(physeq = ex_ps, fasta_df = key_to_seq_map)
ps_renamed
```


```{r, eval = FALSE, echo = FALSE}
#(Optional) If you wanted to test that your new human readable names match exactly to the old ASVs, you could do something like this. 
# Code for checking rename_taxa_with_short_names 
otu_old = as.data.frame(phyloseqCompanion::otu.data.table(ex_ps))
otu_old
otu_new = as.data.frame(phyloseqCompanion::otu.data.table(ps_renamed))
dim(otu_new)
tax_old = as.data.frame(phyloseqCompanion::taxa.data.table(ex_ps))
rownames(tax_old) = tax_old$Taxon
tax_new = as.data.frame(phyloseqCompanion::taxa.data.table(ps_renamed))
rownames(tax_new) = tax_new$Taxon

all_true_otu = 1
all_true_tax = 1
for(i in 1:nrow(key_to_seq_map)){
  cur_id = key_to_seq_map[i,"seq.name"]
  cur_seq = key_to_seq_map[i, "seq.text"]
  if(sum(otu_old[, cur_seq] == otu_new[, cur_id]) == nrow(otu_new)){}else{all_true_otu = 0}
  if(! identical(as.vector(tax_new[cur_id,c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")]), as.vector(tax_old[cur_seq,c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus")]))){all_true_tax =0}
}
if(all_true_otu == 0){print("FAIL: function rename_taxa_with_short_names not working as expected")}else{print("PASS")}
if(all_true_tax == 0){print("FAIL: function rename_taxa_with_short_names not working as expected")}else{print("PASS")}


```


Last, remember how there was one ASV which was poorly aligned? That ASV isn't in our tree but its in our phyloseq object. We will need to remove that. 
```{r, eval = TRUE, echo = TRUE}

# Combine our tree with the phyloseq - remember there is one ASV we need to remove because it was filtered as producing a poor quality alignment.
ps_pruned <- phyloseq::prune_taxa(pruned_tree$tip.label, ps_renamed)
ps_pruned
```

Ok, and now we are finally ready to pass off the tree to the phyloseq object. 
```{r, eval = TRUE, echo = TRUE}
# Last, add the phy_tree to the phyloseq object
ps_with_tree = phyloseq::phyloseq(phyloseq::otu_table(ps_pruned, taxa_are_rows = TRUE), 
                                  phyloseq::tax_table(ps_pruned),
                                  phyloseq::sample_data(ps_pruned),
                                  phyloseq::phy_tree(midpoint.tree))
ps_with_tree
```

Now we can perform our community analyses on it such as UniFrac. Read more about UniFrac [here](https://pmc.ncbi.nlm.nih.gov/articles/PMC1317376/pdf/1021-05.pdf)
```{r, eval = TRUE, echo = TRUE}
# Unweighted UniFrac
unifrac_unweighted = phyloseq::UniFrac(ps_with_tree, weighted = FALSE)
adonis_result_unweighted <- vegan::adonis2(unifrac_unweighted ~ When, 
                                data = as.data.frame(as(phyloseq::sample_data(ps_with_tree), "data.frame")))
print(adonis_result_unweighted)
ordination_unifrac <- phyloseq::ordinate(ps_with_tree, method = "PCoA", distance = "unifrac")
phyloseq::plot_ordination(ps_with_tree, ordination_unifrac, color = "When")


# Weighted UniFrac
unifrac_weighted = phyloseq::UniFrac(ps_with_tree, weighted = TRUE)
adonis_result_weighted <- vegan::adonis2(unifrac_weighted ~ When, 
                                data = as.data.frame(as(phyloseq::sample_data(ps_with_tree), "data.frame")))
print(adonis_result_weighted)
ordination_unifrac <- phyloseq::ordinate(ps_with_tree, method = "PCoA", distance = "wunifrac")
phyloseq::plot_ordination(ps_with_tree, ordination_unifrac, color = "When")


```


